util/process_wrapper/process_wrapper.cc

```diff
@@ -143,24 +142,74 @@ int PW_MAIN(int argc, const CharType* argv[], const CharType* envp[]) {
       for (++i; i < argc; ++i) {
         arguments.push_back(argv[i]);
       }
-      // after we consume all arguments we append the files arguments
-      for (const System::StrType& file_arg : file_arguments) {
-        arguments.push_back(file_arg);
-      }
     } else {
       std::cerr << "process wrapper error: unknow argument \"" << ToUtf8(arg)
                 << "\"." << '\n';
       return -1;
     }
   }
 
+  // Expand tokens into the file arguments (since some paths might embed
+  // `${pwd}` that we need to resolve it from the subsitution mappings).
+  for (System::StrType& file_arg : file_arguments) {
+    ReplaceTokens(file_arg, subst_mappings);
+  }
+
+  // Process the arguments, making sure that we don't have paths to param files
+  // within another param file (that is an arg composed of `@` + a path), which
+  // rustc doesn't support.
+  // (We process them in reverse so that we can append the param files to the
+  // front of `file_arguments` in the right order).
+  for (auto arg_it = arguments.rbegin(); arg_it != arguments.rend(); arg_it++) {
+    System::StrType& arg = *arg_it;
+    ReplaceTokens(arg, subst_mappings);
+
+    // Check whether this is a regular argument or a path to a param file.
+    if (arg.size() > 0 && arg.substr(0, 1) == PW_SYS_STR("@")) {
+      System::StrType path = arg.substr(1);
+
+      // Read the actual params off the param file so that we can 1) extract
+      // potential other param files inside and 2) substitute mappings.
+      System::StrVecType params;
+      if (!ReadFileToArray(path, params)) {
+        std::cerr << "Cannot read param file " << ToUtf8(path) << "\n";
+        return -1;
+      }
+
+      for (auto par_it = params.rbegin(); par_it != params.rend(); par_it++) {
+        System::StrType& param = *par_it;
+        ReplaceTokens(param, subst_mappings);
+
+        // If we find a path to a param file, move it to the top-level file
+        // arguments instead (we cannot have nested param files).
+        if (param.size() > 0 && param.substr(0, 1) == PW_SYS_STR("@")) {
+          file_arguments.insert(file_arguments.begin(), param);
+          params.erase(std::next(par_it).base());
+        }
+      }
+
+      // Create a new param file so that we don't need to edit the original one
+      // that Bazel created.
+      System::StrType new_path = path + PW_SYS_STR(".expanded");
+      if (!WriteArrayToFile(new_path, params)) {
+        std::cerr << "Cannot write params to file " << ToUtf8(new_path) << "\n";
+        return -1;
+      }
+
+      arguments.erase(std::next(arg_it).base());
```

Comment:
Oh this looks complicated. Would using `remove` or `remove_if` (https://en.cppreference.com/w/cpp/algorithm/remove) and https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom after the for loop ends be simpler to reason about?

I know it doesn't matter in our workloads, but even hypothethically repeatedly prepending into a vector makes me nervous. How about collecting new file arguments into a new vec and merging them once after the end of the loop?

Reply:
Yeah that's a fair callout. I'll look at simplifying that.